### PR TITLE
AddrSplosion -- Partie Deux Avec Vengeance

### PR DESCRIPTION
**Pull request by jbenet**
This PR reduces the effects of addr-splosion.

- [x] Nodes clear their own internal cache when they
      receive ground truth from the peer itself.

- [x] Nodes' address cache expires. Each addr has a
      TTL (for now set to 1hr). After which nodes
      must resolve addrs again. (this may be too
      aggressive and can be revisited once we bring
      in proper addr management)

Related to: #573, #603

### CODE REVIEW

File: p2p/peer/peerstore.go [hunk 1]
```diff
@@ -73,35 +93,57 @@ func (ab *addressbook) Addresses(p ID) []ma.Multiaddr {
...
+		if m.Expired() {
+			expired = append(expired, s)
+		} else {
+			good = append(good, m.Addr)
+		}
+	}
+
+	// clean up the expired ones.
+	for _, s := range expired {
+		delete(ab.addrs[p], s)
```
Comment (briantigerchow) [inline p2p/peer/peerstore.go]: @jbenet need to hold write lock to delete these?

Comment (jbenet) [inline p2p/peer/peerstore.go]: yep, it is. defered. github's omitting it from the diff. (but would've been a great catch)

Comment (briantigerchow) [inline p2p/peer/peerstore.go]: I meant that the lock is currently held for read access.

```Go
ab.RLock()
defer ab.RUnlock()
```

File: p2p/peer/peerstore.go [hunk 2]
```diff
@@ -38,20 +44,34 @@ type Peerstore interface {
...
+	return time.Now().After(e.TTL)
 }
 
-type addressMap map[string]ma.Multiaddr
+type addressMap map[string]expiringAddr
 
 type addressbook struct {
 	addrs map[ID]addressMap
+	ttl   time.Duration // initial ttl
 	sync.RWMutex
```
Comment (briantigerchow) [inline p2p/peer/peerstore.go]: mu over the things it locks?

File: p2p/peer/peerstore.go (hunk 1 above)
Comment (briantigerchow) [inline p2p/peer/peerstore.go]: but needs to be `Lock()` `Unlock()` since there exist map deletions.

Comment (jbenet) [inline p2p/peer/peerstore.go]: Ahhhh great catch. right you are.

File: p2p/peer/peerstore.go (hunk 2 above)
Comment (jbenet) [inline p2p/peer/peerstore.go]: sure, i'm down to use that convention. i think there should be a comment declaring what it guards:

```go
type foo struct {
  amu // guards a
  a

  bmu // guards b
  b
}
```

Comment (briantigerchow) [inline p2p/peer/peerstore.go]: SGTM

**Comment by briantigerchow** [comment]
LGTM

**Pull request merged**